### ISSUE TITLE
a quest that is not desired by the majority of a party is too hard to remove

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#5560]
It's possible for a party to contain a quest that the majority of the members don't want. E.g., the quest owner uses a scroll and then immediately starts the quest for purely their own benefit with no other participants, or if the quest owner doesn't log in for many days after sending invitations (a single inactive party member can then prevent the quest from being started). The workarounds currently are for party members to leave the party and start a new one (troublesome and often undesirable) or for an admin to delete the quest (not hard, but we're reluctant to do that prematurely, resulting in further delays for the party).

I'm creating this issue because @hairlessbear has outlined some interesting and promising ideas to me and I want to hear more from him about it. Even if we end up not going with his ideas, some solution would be desirable.
##

**Comment by hairlessbear** [comment]
Thanks @Alys! 

Preface: whenever I say "Abort" in the following, I'm referring to both Cancel and Abort. As in, all of the ideas below would apply when a quest is pending ("Cancel") and when a quest is in progress ("Abort").

**TL;DR**: The idea is to use a voting mechanism, allowing the party to collectively abort a quest. I'd like to add a "Vote to Abort" button next to Ian for all users except the quest leader (it would be in the same place that Abort button goes for the quest leader). If a majority of users click the button, then the quest is cancelled, simple as that. Once clicked, the button would become disabled and contain text like "You've already voted". It may also be prudent to add a "Vote to Cancel" button to the quest invite modal.

More ideas (these are just initial thoughts):
1. If a quest is aborted in this way, the behavior shouldn't be identical to what happens when a quest leader aborts a quest. Since this should only happen when something has presumably gone wrong, some additional actions should occur. Some ideas:
   1. Send an automated message to the quest leader informing them that their party has aborted their quest, and suggesting that they may want to talk to their party about this.
   2. An option (or requirement?) for everyone casting a vote to give a reason for their vote, which is sent to the quest leader (the sender would be anonymized)
   3. Prevent a user who had a quest cancelled in this manner from starting another quest for 24 hours
      1. If a user has multiple quests cancelled in a row (or over a preset amount of time), extend the length of time they are prevented from starting quests. This same method could possibly be used to boot a user from a party, in extreme circumstances (e.g., the last 10 quests they started were all aborted before the quest began)
2. It may be useful to display the current vote count. This could allow users to learn that they aren't the only ones who feel negatively about a quest, which could encourage them to vote when they otherwise wouldn't have. On the other hand, it could be abused. For example, a user could be obnoxious by casting a vote on every quest, and no one would know who is doing it.
3. It might be useful for the quest leader to have non-anonymized voting data available to them. This would allow them to weed out anyone abusing the system, and take appropriate action.
4. Voting shouldn't be used to start a quest. If the quest leader is inactive, this would force an inactive user to participate in a quest, which is problematic for multiple reasons.

**Comment by hairlessbear** [comment]
One other idea that's not voting related.

We could add an abort button for all users after a quest leader has been inactive for a preset amount of time. So if they haven't logged in for, say, 3 days, any individual can abort the quest. I strongly prefer voting, but this is another option.

**Comment by Alys** [comment]
Some of my comments here come from my feeling that a party should be a collaborative group that can solve at least some of its own problems with communication, or failing that, by removing recalcitrant members (or removing yourself if the party doesn't suit you). This voting mechanism would certainly help with that! I'm also pretty keen to see something like this done, so I'm commenting on things that I think could be done in a second stage after the basics are in place. Note that my comments are not meant to be "this is how it will be done"; they're just my current opinions. :)

_"If a majority of users click the [vote to cancel/abort] button"_  -- Yes, where "majority" is calculated by not counting the quest leader or any member who has been inactive for X number of weeks. X=1?

_"Once clicked, the button would become disabled"_ -- Or we could allow it to be unclicked to remove your vote. However that could be added in a second stage.

1 ii _"... give a reason for their vote ..."_ -- I'm thinking that's extra complexity that might not be needed, at least in the initial implementation (maybe add in a second/third stage). In small parties, the reasons would be less anonymous. Non-anonymous communication might be better for building a strong, honest party. It it were not anonymous, a PM or chat message would suffice. I'm not against this, I just don't think it's one of the most important requirements.

1 iii _"Prevent a user who had a quest cancelled in this manner from starting another quest for 24 hours"_ -- For simplicity, this could be implemented as a buff that wears off on cron, which I know can be much less than 24 hours, but SO simple. :)  Although really, setting a time and then checking that when the user clicks on a quest scroll wouldn't be that hard.

1 iii a _"If a user has multiple quests cancelled in a row ..."_ -- I don't think we should do anything automatic here. If the user keeps on messing up in the same way, the party should boot them (or leave the party if they are the leader). Letting the party handle this would be MUCH simpler than keeping track of historical information (fewer chances of bugs too).

2 _"It may be useful to display the current vote count..."_ -- I'd be in favour of a vote indicator next to each person's name, so that voting isn't anonymous. That would help with the obnoxious user problem, and allow everyone to know how everyone is feeling. Currently, opinions about current/future quests have to be expressed in party chat, so non-anonymous voting isn't that much different. 

@hairlessbear You might want to let us know your opinions about anonymity, as a contrast to mine.

**Comment by hairlessbear** [comment]
"_where "majority" is calculated by not counting the quest leader or_ **_any member who has been inactive for X number of weeks. X=1?**_" -- how would the bolded part work exactly? Would they just not have a button? Would the button be disabled with a special message (e.g., "you may not vote because you've been inactive for X weeks")? 

"_...Or we could allow it to be unclicked to remove your vote._" :+1: :+1: :+1:  I wish I would've thought of that. I think this should be a requirement of the first implementation. If the button goes where I'm suggesting, it will be right next to the Accept/Reject buttons added in [#5298](https://github.com/HabitRPG/habitrpg/pull/5298#issuecomment-111891025), making accidental votes very possible.

When it comes to anonymity, I'm divided. I'm 100% in favor of strong communication and honesty among party members, and agree that non-anoymized votes could improve both of these. I do like the idea of everyone knowing who voted for what, especially if it prompted a discussion about what quest should be done, and why etc. I could even see scenarios in which non-anonymized voting caused conversations to occur that wouldn't have otherwise.

That being said, I'm worried that fear of embarrassment/shame will prevent users from voting if there's no anonymity. In my anecdotal experience, most people try to avoid conflict as much as possible, and definitely don't want to be the first one to speak in opposition to almost any topic. So, if the voting wasn't anonymous, I'm worried that few people would use it, and that they would be even less likely to use it in particularly uncomfortable scenarios, when it could be most useful. I think this is a particularly strong risk in this context. Since quests are, in the grand scheme of things, not important, it's extremely easy to rationalize going along with one even if you don't want to. And since most users in most situations would likely be unaware of their fellow party members' opinions, I think they'd be even more likely to stay quiet.

I can think of strong arguments for both. I'm more in favor of anonymity, but I'll definitely understand if you end up choosing non-anonymous voting.

**Comment by crookedneighbor** [comment]
Just my 2 cents, but I'd prefer voting in parties to not be anonymous. 

@hairlessbear regarding the inactive user thing, I believe what Alys was trying to say is that when calculating a majority vote, you wouldn't need to count users that had been inactive for a week or longer. 

For instance, say you had a group of 6 people. All members actively log in. Not counting the quest starter, you'd have 5 active members and would need at least 3 + or - votes to constitute a majority.

Now let's say you have a group of 6, where 2 members have not logged in for a month. Take out the quest starter and that gives you 3 active members. You'd only need 2 + or - votes.

If one or both of the previously inactive members logged in, they would become active members again and they would count towards the majority. No need to hide the button at all.

**Comment by hairlessbear** [comment]
@crookedneighbor: Ah, that makes sense, thanks!

After thinking about it further, I'm on board for non-anonymous voting. Even if the issues I outlined occur sometimes, I think that non-anonymous votes are more likely to be beneficial than not (and it also fits better with HabitRPG's community and philosophy). 

EDIT: Not only that, but I originally didn't stop to consider the _downsides_ to anonymized voting, like its potential to breed mistrust among party members. I'm definitely all for non-anonymized votes.

**Comment by LittleMupsMum** [comment]
As one of the users who had this issue, I'm not sure your fixes address the situation we had.  Either that or I've misunderstood the fix.
We had a party member send out invitations unexpectedly (since someone else was going to start a different quest), then not log in for over a week.  It was an easy quest which we thought we'd be able to complete quickly, so those of us who were active accepted the quest, and we had no way of knowing the quest owner would disappear.  So what we wanted was not a way to end the quest, but a way to start it without needing the now inactive quest owner to do it.  The voting system you suggest would give us a way to end the quest, but would that require for us to have not accepted it?  Or would we still be able to vote to end the quest even after accepting it?

Also, in the case that a party does vote to end a quest, what happens to the quest scroll?  Our quest leader took a (unilateral) decision to boot the quest owner from the party to see if it would end the quest (and also reinvite the owner immediately), but the actual result was that the quest remained with a new owner who was then able to start it.  In hindsight this wasn't the fairest thing for the quest owner as they then lost their quest scroll, without being able to complete the quest.  In a vote-to-abort situation, would the quest scroll go back to its owner?

**Comment by hairlessbear** [comment]
@LittleMupsMum: Thanks for your input! On starting the quest via voting, point 4 in my original post is "_Voting shouldn't be used to start a quest. If the quest leader is inactive, this would force an inactive user to participate in a quest, which is problematic for multiple reasons._" To expand on this, here are some example scenarios that illustrate why I believe this would be problematic:
1. A user with a lot of dailies sends out a quest, then goes inactive, but does not check in to the Inn. If the party starts the quest without them, and the user comes back before the quest is over, they would do a huge amount of damage to all party members.
2. A user sends out a quest invitation then goes inactive. The party votes to start the quest without them, and completes it before they return. It's possible that the quest owner would be frustrated by this, because although they would get the rewards from their quest, they didn't get to actually participate in it. Additionally, in most cases, the quest would have been started without their knowledge (after all, they're inactive, so it's unlikely they're keeping up-to-date with Habitica). Both of these factors could lead to a quest leader getting frustrated with a party, potentially causing conflict.

On when you're allowed to cancel: the voting system would always allow you to cancel the quest, even after you've accepted the invitation. As long as the majority of active party members vote to cancel, it would be canceled. This could even take place after a quest has started, even if you're not a participant of the quest (this is for instances where a user sends a quest invitation and then starts it immediately so they're the only participating member).

Finally, yes, in a vote-to-abort situation, the quest scroll would go back to its owner. The behavior would be identical to if the quest leader aborted the quest themselves.

**Comment by Alys** [comment]
Party leaders are able to abort quests now, as well as quest owners, so there's less chance of a quest being stuck. The ideas listed here a good but more complex than possible for a quick fix, and while the website is being improved and expanded in various ways, it's probably best to avoid complex additions that aren't essential. I'm closing this but we can still refer back to it when needed.
